Candidate patches

titu1994 opened the pull request:
# Bugfixes
- Correctly call `multi_test_epoch_end()` when performing `trainer.test()`.
- Raise warning per call instead of raising NotImplementedError() when `multi_*` methods are not implemented but subclass also does not override `validation_epoch_end()` and `test_epoch_end()`. This is an uncertain state between unknowingly supporting multi data loaders but not overriding any of the required methods.
- Add min_lr to all NeMo schedulers so that min_lr is properly preserved for all schedulers.
- Patch schedulers to properly support target / cls instead of `name`. 
- Schedulers now fail with warning that they were not initialized instead of raising ValueError. This is necessary when re-instantiating models from configs but data loaders are not available to compute effective `max_steps`.
- Optimizers now preserve the config passed to them via deepcopy.

# Improvements
- Add helper method to correctly test a model on the test set.
- Config check when optimizer parameters are passed via their Param classes.
- Add full test suite for Optimizers and Schedulers

okuchaiev commented on nemo/core/classes/modelPT.py:
```diff
@@ -583,19 +583,79 @@ def test_epoch_end(
     def get_test_dataloader_prefix(self, dataloader_idx=0):
         return self._test_names[dataloader_idx]
 
+    def prepare_test(self) -> bool:
```
is this patch around PTL?

titu1994 replied:
The optimization fix is over AMP (I'm looking for a better way to solve this), and the DDP one is PTL

blisc commented on nemo/core/config/optimizers.py:
```diff
@@ -238,6 +240,11 @@ def get_optimizer_config(name: str, **kwargs: Optional[Dict[str, Any]]) -> Optim
 
     scheduler_params = AVAILABLE_OPTIMIZER_PARAMS[name]
+
+    if kwargs is not None and len(kwargs) != 0:
+        kwargs = OmegaConf.create(kwargs)
+        OmegaConf.merge(scheduler_params(), kwargs)
```
I've been thinking more about this construction. We honestly could do away with dataclasses if we just use `hydra.util.instantiate` instead and pass normal python arugments as opposed to DictConfigs to our classes.

titu1994 replied:
Instantiate requires the assumption you are following the `target` and `params` pattern which causes very deep nest when trying to override parameters for optimizers - `model.optim.sched.params.weight_decay` which is untenable.

blisc replied:
This exposes an issue with the construction of our yaml files, not a problem with hydra.
Nothing is stopping us from exposing it at `sched.params.weight_decay`, or `sched.weight_decay`, or even `weight_decay`.

titu1994 replied:
Well we need to make certain assumptions regarding this, and using classpath everywhere is problematic in all cases other than pure productions configs (which we won't even need there since we will simply be using config objects) so for research flexibility the current schema is practical and useful enough.

blisc commented on nemo/core/config/optimizers.py:
```diff
@@ -238,6 +240,11 @@ def get_optimizer_config(name: str, **kwargs: Optional[Dict[str, Any]]) -> Optim
         )
 
     scheduler_params = AVAILABLE_OPTIMIZER_PARAMS[name]
+
+    if kwargs is not None and len(kwargs) != 0:
+        kwargs = OmegaConf.create(kwargs)
+        OmegaConf.merge(scheduler_params(), kwargs)
+
     scheduler_params = partial(scheduler_params, **kwargs)
```
You can avoid the partial if you want
```python
    scheduler_params = OmegaConf.structured(AVAILABLE_OPTIMIZER_PARAMS[name])
    if kwargs is not None and len(kwargs) != 0:
        kwargs = OmegaConf.create(kwargs)
        scheduler_params = OmegaConf.merge(scheduler_params, kwargs)
    return scheduler_params  # A DictConfig not a (class pointer/definition, not sure what to call this)
```

titu1994 replied:
Scheduler parameter dataclasses are somewhat special, they dont hold the `max_steps` kwarg even though their constructor does have this value (in most but not all constructors). I have to defer its calculation until we resolve either `iters_per_batch` or the actual passed `max_steps` and finally instantiate the constructor of the scheduler with the union of the params and this `max_steps`.

titu1994 commented:
@blisc @okuchaiev I've simplified optimizer construction, so no more need to explicitly call `setup_optimization()` inside any constructor. This has the additional benefit that we dont need to have edge cases for AMP during testing.

Also added a patch for MFCC after recent tts PR which altered signature of `get_seq_length()` to account for MFCC signature. (It doesnt get caught in MatchboxNet tests since torchaudio is not installed on CI so I force it to switch to MelSpec).

omry commented on nemo/core/optim/lr_scheduler.py:
```diff
@@ -434,17 +430,32 @@ def prepare_lr_scheduler(
-
     # Try instantiation of scheduler params from config class path
     try:
-        scheduler_conf = hydra.utils.instantiate(scheduler_args)
```
You replace one line of instantiate with many lines which are more complicated.
I am curious as to the motivation here.
